enhancements/olm/marketplace-api-deprecation.md

```diff
@@ -0,0 +1,107 @@
+---
+title: marketplace-api-deprecation
+authors:
+  - "@ecordell"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-01-06
+last-updated: 2020-01-06
+status: implementable
+---
+
+# Marketplace Operator API Deprecation
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement covers the remaining steps needed to ensure that everything the marketplace-operator APIs have been providing is covered by other features within OLM or OpenShift.
+
+## Open Questions
+
+1. Do we wish to block an upgrade of a cluster if we can detect OperatorSource/CatalogSourceConfig usage that would be affected by the deprecation? (i.e. non-default catalogs using these features).
```

Comment:
The other option is to delete them but retain the CatalogSources generated by them (removing the ownerrefs). This would make existing catalogs continue to work but not receive updates from appregistry, and would push a user to look at them and follow a migration guide.

Another option is like you say - remove the API but migrate the CatalogSources to an "appregistry" type, which would be straightforward to implement but doesn't achieve our goal of moving users off of appregistry as a backend.

Reply:
I prefer blocking the upgrade, as it explicitly requires users to acknowledge and address the deprecation. I recognize that this approach is more disruptive, but I believe that the benefits of an informed cluster admin outweighs the cost.

Reply:
I lean towards @gallettilance's suggestion of alerting on OperatorSources that are not default (though we need to double check that those alerts wouldn't also potentially block an upgrade, I seem to recall a bug in openshift-monitoring where firing alerts would block).

We also currently don't have a way to block minor upgrades but not patch upgrades (there's a proposal to change this but it is not finalized). This means that if we set `Upgradable: false` due to non-default OperatorSources in 4.4, we would prevent the 4.4.1 update even though we really would only want to block 4.4->4.5 when OperatorSources are actually deprecated.

Reply:
> I seem to recall a bug in openshift-monitoring where firing alerts would block

@ecordell I believe that you and I looked into this earlier and that it only affects e2e tests, not upgrades.

Reply:
New information: the proposal has been confirmed, and in 4.4 upgradeable: false will block Y stream updates only. I think that makes blocking updates the most reasonable choice - will update.